cmd/gke-certificates-controller/app/gke_certificates_controller.go

```diff
@@ -94,9 +105,64 @@ func Run(s *GKECertificatesController) error {
 		signer.handle,
 	)
 
-	sharedInformers.Start(nil)
-	// controller.Run calls block forever.
-	go approveController.Run(5, nil)
-	go signController.Run(5, nil)
-	select {} // block forever
+	run := func(stopCh <-chan struct{}) {
+		sharedInformers.Start(stopCh)
+		// controller.Run calls block forever.
+		go approveController.Run(5, stopCh)
+		go signController.Run(5, stopCh)
+		<-stopCh
+	}
+
+	if s.LeaderElectionConfig.LeaderElect {
+		leaderElectionClient, err := clientset.NewForConfig(restclient.AddUserAgent(kubeconfig, "leader-election"))
+		if err != nil {
+			return err
+		}
+		leaderElectionConfig, err := makeLeaderElectionConfig(s.LeaderElectionConfig, leaderElectionClient, recorder)
+		if err != nil {
+			return err
+		}
+		leaderElectionConfig.Callbacks = leaderelection.LeaderCallbacks{
+			OnStartedLeading: run,
+			OnStoppedLeading: func() {
+				utilruntime.HandleError(fmt.Errorf("lost master"))
+			},
+		}
+
+		leaderElector, err := leaderelection.NewLeaderElector(*leaderElectionConfig)
+		if err != nil {
+			return err
+		}
+		leaderElector.Run()
+		return fmt.Errorf("lost lease")
```

Comment:
Does this happen when mastership is lost or on some unrecoverable internal error?

Reply:
It should never reach this point as leaderElector.Run() calls OnStoppedLeading before returning which should kill the process.

Reply:
Can you make it a `panic("unreachable")` ?

Reply:
Done :)